Template files fix

mooreryan opened the pull request:
Work in progress on fixes for `template-files` branch (there are still some bugs after the changes here--needs a bit more).  I left some comments to try to explain whats going on.

mvolkmann committed:
attempt to move templates to separate files

mooreryan committed:
Move dog type into its own module

This is one way to break the cyclic dependencies that were present between the modules.

mooreryan commented on dream_demo/bin/dog.ml:
```diff
@@ -0,0 +1,8 @@
+open Ppx_yojson_conv_lib.Yojson_conv.Primitives
+
+type t = { id : string; name : string; breed : string }
```
I changed the type name from `dog` to `t`.  It is customary to have the main type of a module be called `t`.

mooreryan commented on dream_demo/bin/dog.ml:
```diff
@@ -0,0 +1,8 @@
+type t = { id : string; name : string; breed : string }
+[@@deriving yojson]
+
+let breed = Option.fold ~some:(fun dog -> dog.breed) ~none:""
```
About the `breed` and `name` functions: I would expect `Dog.breed` to take `t` rather than `t option` (same with `name`).  May be a good idea to name these a bit differently to signal it takes `t option`. Potentially `breed'`.

mvolkmann replied:
I renamed lots of things to end in `_opt`. Hopefully it is more clear now.

mooreryan commented on dream_demo/bin/dog_row.eml.html:
```diff
@@ -0,0 +1,27 @@
+let render dog attrs =
+<tr class="on-hover" id="row-<%s dog.Dog.id%>" <%s attrs %>
```
Need to specify the `Dog` module here.

mvolkmann replied:
Can you explain why that is necessary? I see that it doesn't work without that, but I don't understand why. It isn't needed in `dog.name` and `dog.breed`. What is special about the `id` field?

mooreryan commented on dream_demo/bin/dune:
```diff
@@ -8,4 +8,14 @@
 (rule
   (targets main.ml)
   (deps main.eml.ml)
+  (action (run dream_eml %{deps} --workspace %{workspace_root})))
```
Sorry about the auto-formatting that was applied to this file. Nothing was changed in this file though.

mooreryan commented on dream_demo/bin/form.eml.html:
```diff
@@ -0,0 +1,21 @@
+  > <%s! Dream.csrf_tag request %>
+    <div>
+      <label for="name">Name</label>
+      <input id="name" name="name" required size="30" type="text" value="<%s Dog.name selected_dog_option %>" />
```
`dog_name` -> `Dog.name`

mooreryan commented on dream_demo/bin/form.eml.html:
```diff
@@ -0,0 +1,21 @@
+    </div>
+    <div>
+      <label for="breed">Breed</label>
+      <input id="breed" name="breed" required size="30" type="text" value="<%s Dog.breed selected_dog_option %>" />
```
`dog_breed` -> `Dog.breed`

mooreryan commented on dream_demo/bin/form.eml.html:
```diff
@@ -0,0 +1,21 @@
+let render request attrs selected_dog_option =
```
Same comment about the auto-formatting -- it obscures the actual change, sorry about that.

mooreryan commented on dream_demo/bin/main.eml.ml:
```diff
@@ -15,90 +12,10 @@ let generate_uuid () = Uuidm.(v `V4 |> to_string)
 let add_dog name breed =
   let id = generate_uuid () in
-  let dog = { id; name; breed } in
+  let dog = { Dog.id; name; breed } in
```
Need to specify module here as well.

mvolkmann replied:
Why is that needed on the `id` field, but not on `name` or `breed`?

mvolkmann commented:
@mooreryan I love all the changes you are making so the code is more idiomatic! As a newcomer to OCaml, I really need that kind of advice. Let me know when you are finished making changes.

mooreryan committed:
Remove param function

The "/form" route doesn't have a path parameter, so I'm not sure what Dream.param is attempting to do here--it will raise an error for missing the parameter.

mooreryan committed:
Specify module

mooreryan committed:
Fix dog rendering for rows

The "/table-rows" route doesn't have a path parameter, so I'm not sure what Dream.param is attempting to do here--it will raise an error for missing the parameter.

mooreryan committed:
Specify module

mooreryan committed:
Fix templating

The auto-formatting is messing up the templates.  The `%` characters by themselves to specify OCaml code that shouldn't be included in the html need to be in the first column. The auto-formatting keeps preventing that.  So I switched to the alternative syntax.  Unfortunately, the auto-formatting of the html will not align the tags properly....

mooreryan commented:
Added a few more commits--the git commit descriptions have some comments you may want to look at. Specifically regarding the use of `Dream.param`.  It may be that I was misinterpreting your intent with that.

mooreryan commented:
With regards to code changes to make things more idiomatic...mainly I was changing things the minimum amount to get it working. There's probably more things you could change that could make it feel more natural, but tbh I wasn't focused on that in this PR.

But, the compilation errors are all fixed and the site seems to work when I'm testing it.  You should try it out and make sure I didn't break any critical behaviors.

Otherwise this PR could be considered done in terms of just getting things to compile and run.

mvolkmann commented:
Thanks so much Ryan! I'll check it out.
